docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,392 @@
+        ]
+
+Types
+-----
+
+A type is a string that identifies a set of values.
+In each of the extension points mentioned above, the first column is always the name of a type.
+A type can be defined by adding **typeModel** tuples for that type. Additionally, the following built-in types are available:
+
+- The name of an NPM package matches imports of that package. For example, the type **express** matches the expression **require("express")**. If the package name includes dots, it must be surrounded by single quotes, such as in **'lodash.escape'**.
+- The type **global** identifies the global object, also known as **window**. In JavaScript, global variables are properties of the global object, so global variables can be identified using this type. (This type also matches imports of the NPM package named **global**, which is a package that happens to export the global object.)
+- A qualified type name of form **<package>.<type>** identifies expressions of type **<type>** from **<package>**. For example, **mysql.Connection** identifies expression of type **Connection** from the **mysql** package. Note that this only works if type annotations are present in the codebase, or if sufficient **typeModel** tuples have been provided for that type.
+
+Access paths
+------------
+
+The **path**, **input**, and **output** columns consist of a **.**-separated list of components, which is evaluted from left to right, with each step selecting a new set of values derived from the previous set of values.
+
+The following components are supported:
+
+- **Argument[**\ `number`\ **]** selects the argument at the given index.
+- **Argument[this]** selects the receiver of a method call.
+- **Parameter[**\ `number`\ **]** selects the parameter at the given index.
+- **Parameter[this]** selects the **this** parameter of a function.
+- **ReturnValue** selects the return value of a function or call.
+- **Member[**\ `name`\ **]** selects the property with the given name.
+- **AnyMember** selects any property regardless of name.
+- **ArrayElement** selects an element of an array.
+- **Element** selects an element of an array, iterator, or set object.
+- **MapValue** selects a value of a map object.
+- **Awaited** selects the value of a promise.
+- **Instance** selects instances of a class.
+
+The following components are called "call site filters". They select a subset of the previously-selected calls, if the call fits certain criteria:
+
+- **WithArity[**\ `number`\ **]** selects the subset of calls that have the given number of arguments.
+- **WithStringArgument[**\ `number`\ **=**\ `value`\ **]** selects the subset of calls where the argument at the given index is a string literal with the given value.
+
+Components related to decorators:
+
+- **DecoratedClass** selects a class that has the current value as a decorator. For example, **Member[Component].DecoratedClass** selects any class that is decorated with **@Component**.
```

Comment:
Since decorators are uncommon, it would be nice to see an example of this.

Reply:
On the other hand, decorators are uncommon, so the reader most likely won't care. I think if you come to the document motivated by an example that involves decorators, it wouldn't be that hard to figure out how to use these.